### PR TITLE
SNOW-3031757: Add more SQL Binary to C tests

### PR DESCRIPTION
**Pull request by sfc-gh-pbulawa**
## Add comprehensive test coverage for BINARY type conversions

This pull request enhances the ODBC test suite with extensive coverage for BINARY data type conversions, focusing on edge cases and incompatible type conversions.

### New Test File
- **`binary_incompatible_conversions.cpp`**: Tests all incompatible C type conversions for BINARY data, ensuring proper SQLSTATE 07006 errors are returned for integer, floating-point, numeric, temporal, interval, and GUID C types.

### Enhanced Binary to SQL_C_BINARY Tests
- Added VARBINARY synonym conversion test
- Added exact-fit buffer handling test
- Added 3-chunk retrieval test for large binary data
- Added zero-length buffer test for length-only queries

### Enhanced Binary to Character Type Tests  
- Added VARBINARY to SQL_C_CHAR and SQL_C_WCHAR conversion tests
- Added exact-fit buffer tests for both SQL_C_CHAR and SQL_C_WCHAR
- Added truncation edge case tests (one-byte-short buffer, buffer size 1)
- Added 3-chunk retrieval test for hex string output

### Feature File Updates
- Updated `binary_conversion_to_c_binary.feature` with new test scenarios
- Updated `binary_conversion_to_c_char.feature` with additional edge cases
- Added `binary_incompatible_conversions.feature` documenting all incompatible conversion scenarios

### CODE REVIEW

**Comment by sfc-gh-pbulawa** [comment]
* **#718** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/718?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/718?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: odbc_tests/tests/e2e/types/binary_conversion_to_c_char.cpp [hunk 1]
```diff
@@ -237,3 +259,131 @@ TEST_CASE("should retrieve large binary as hex in chunks via SQLGetData with SQL
...
+  const auto stmt = conn.execute_fetch("SELECT X'ABCDEF'::BINARY");
+
+  // Then SQL_C_CHAR with buffer = 6 (one short of the 7 needed) should return 01004
+  char buffer[6] = {};
+  SQLLEN indicator = 0;
+  SQLRETURN ret = SQLGetData(stmt.getHandle(), 1, SQL_C_CHAR, buffer, 6, &indicator);
+  REQUIRE(ret == SQL_SUCCESS_WITH_INFO);
+  REQUIRE(get_sqlstate(stmt) == "01004");
+  REQUIRE(indicator == 6);
+}
```
[the remark this answers is not in the document]
Comment (sfc-gh-pbulawa) [reply odbc_tests/tests/e2e/types/binary_conversion_to_c_char.cpp:343]: Added check.

**Review by sfc-gh-makowalski — APPROVED** [review]
